feat: disable background throttling

bastiankistner opened the pull request:
This PR relates to https://github.com/tauri-apps/tauri/issues/5250

It requires a change of tauri for which I've opened another PR: https://github.com/tauri-apps/tauri/pull/12181

I've added an option `disableBackgroundThrottling` to allow developers to choose if their view should automatically be throttled / suspended by the browser after a certain time of being hidden or minimised. In the worst case, this default behaviour of browsers can lead to broken applications running in the background. 

For instance if a websocket connection is the only reason not to suspend a "tab" and the machine goes into sleep mode and wakes up again, the connection is often lost and the browser won't have any reason to keep the background tab active and therefore unload it, which will not allow the hidden view to "heal" itself. 

This default behaviour effectively renders any browser capabilities unusable for long running background processes. One alternative might be moving logic from the frontend to the backend, but browsers have meanwhile become a great choice for local first apps and if the browser already offers the required features easily accessible, why force developers to leave a territory they feel comfortable with?

As of now, this is only available for WebKit based browsers on iOS 17.0+,iPadOS 17.0+,Mac Catalyst 17.0+, macOS 14.0+, visionOS 1.0+ ( see https://developer.apple.com/documentation/webkit/wkpreferences/inactiveschedulingpolicy-swift.enum?language=objc ).

It can be assumed that it will also be implemented for WebView2 (Windows). And it might be that it's already usable for linux and android, although I couldn't find any documentation about it and also cannot test it locally due to a lack of devices. 

As an alternative for windows and non-webkit browsers, one can setup a pending WebLock transaction, which, according to the docs, can also keep the browser from suspending it.

----

I needed to update a snapshot test and am unsure if this is at all related to this change. Please let me know if I did something wrong here.

bastiankistner committed:
feat: disable background throttling

webkit only for now

bastiankistner committed:
chore: changes file

pewsheen commented on src/wkwebview/mod.rs:
```diff
@@ -347,6 +347,18 @@ impl InnerWebView {
 
+      // disable background throttling if attributes.disable_background_throttling is true
+      // which works for iOS 17.0+,iPadOS 17.0+,Mac Catalyst 17.0+, macOS 14.0+, visionOS 1.0+
+      if attributes.disable_background_throttling {
```
We may need to add a version check before setting the preference value in case an unsupported version will panic.

Reference:
https://github.com/tauri-apps/wry/blob/2ffcd494af480ddbade3d7568e4d133ba519d83f/src/wkwebview/mod.rs#L195

bastiankistner replied:
Will look into that tomorrow! I assumed that if there is an unsupported version, it'll just be ignored. But that was just a (bad) guess. Thank you very much @pewsheen !

FabianLars replied:
Sometimes that's the case, sometimes not 😂 In this case i would assume it'll be ignored as well. I have a vm with something older than 14 that i could check tomorrow.

pewsheen replied:
And I think the attribute will affect iOS as well, so we might need to validate it on iOS, too 👀

FabianLars replied:
@pewsheen was correct. Without a guard it panics on macos 12

bastiankistner replied:
@FabianLars thank you for testing! I've added a version check

bastiankistner committed:
fix: add version check for background throttling

FabianLars commented on src/lib.rs:
```diff
@@ -583,7 +583,9 @@ pub struct WebViewAttributes<'a> {
   ///
-  /// - **Linux / Windows / Android**: Unsupported yet. But workarounds like a pending WebLock transaction might suffice.
+  /// - **Linux / Windows / Android**: Unsupported. Workarounds like a pending WebLock transaction might suffice.
+  /// - **iOS / macOS**: Supported since version 17.0+.
```
wrong version for macos

FabianLars commented on src/lib.rs:
```diff
@@ -1202,7 +1204,9 @@ impl<'a> WebViewBuilder<'a> {
   ///
-  /// - **Linux / Windows / Android**: Unsupported yet. But workarounds like a pending WebLock transaction might suffice.
+  /// - **Linux / Windows / Android**: Unsupported. Workarounds like a pending WebLock transaction might suffice.
+  /// - **iOS / macOS**: Supported since version 17.0+.
```
same

bastiankistner replied:
My bad, apologies. Also need to change that in tauri. Thanks a lot.

FabianLars commented on src/wkwebview/mod.rs:
```diff
@@ -349,14 +349,20 @@ impl InnerWebView {
-        );
+      #[cfg(any(target_os = "ios", target_os = "macos"))]
+      {
+        let is_supported_os = (cfg!(target_os = "ios") && os_major_version >= 17)
```
we _could_ replace `is_supported_os` with `custom_data_store_available` from above since that has the same version requirement but i'm not sure if that's good for visibility 🤔

bastiankistner replied:
I can do so, but indeed, that could leave the impression that it has something to do with the custom data store.

FabianLars reviewed (commented):
just a minor doc thing left

bastiankistner committed:
fix: comments

FabianLars commented on .changes/disable-background-throttling.md:
```diff
@@ -0,0 +1,5 @@
+---
+'wry': 'minor:enhance'
```
ah sorry, last one. this can be a patch because wry is still 0.x

FabianLars committed:
Update disable-background-throttling.md

pewsheen commented on src/lib.rs:
```diff
@@ -574,6 +574,21 @@ pub struct WebViewAttributes<'a> {
+  /// - **macOS**: Supported since version 14.0+.
+  ///
+  /// see https://github.com/tauri-apps/tauri/issues/5250#issuecomment-2569380578
+  pub disable_background_throttling: bool,
```
I'm thinking if we should use an enum such as `BackgroundThrottlePolicy` here, just like: https://developer.apple.com/documentation/webkit/wkpreferences/inactiveschedulingpolicy-swift.enum

It would give more flexibility to the developers.

And this property could be renamed to something like `background_throttling`

pewsheen reviewed (commented):
Sorry for the late question, but I think it would be nice to give it more flexibility.

Because it defaults to `suspend` instead of `throttling` on macOS, there might be developers who want to set it to `throttling`. Also, if we want to support other platforms in the future, it may require a breaking change to this property to support different policies.

pewsheen commented:
Related issue: https://github.com/tauri-apps/wry/issues/1246

bastiankistner committed:
feat: use policy enum for background throttling

bastiankistner committed:
Merge branch 'feat/background-throttling-pr' of github.com:bastiankistner/wry into feat/background-throttling-pr

bastiankistner commented:
@pewsheen @FabianLars I changed it to use a policy enum of 
- `disabled`
- `throttle`
- `suspend`

I tested it successfully with my app and it seems to work. But I had to add the dependency `wry` to tauri-utils to implement a From trait. I'm not very experienced with rust, it's been a while... So please let me know if there is anything I should improve.

I also updated the tauri PR.

FabianLars commented on src/lib.rs:
```diff
@@ -1997,6 +2033,17 @@ pub enum PageLoadEvent {
 
+/// Background throttling policy
+#[derive(Debug, Clone)]
+pub enum BackgroundThrottlingPolicy {
```
@pewsheen Should this be non_exhaustive from the start if future compat was the main concern for the enum?

pewsheen replied:
It should be okay for now, as long as the client won't build fail after we add policies.
And Wry will build fail if we forget to handle the new policy.

bastiankistner replied:
Or we go back to the boolean `disableBackgroundThrottling` , [like electron does](https://www.electronjs.org/docs/latest/api/browser-window#:~:text=8859%2D1.-,backgroundThrottling,-boolean%20(optional)%20%2D%20Whether). And maybe add a `backgroundThrottlingPolicy` later that could be used for advanced cases. But then there'd be a dependency of the policy on the disable flag.

I would assume that almost no developer would decide for a semi-throttled/semi-unthrottled mode. In that case, they should rather focus on optimising the usage of timers in their code and keep it fully disabled.

pewsheen replied:
> Or we go back to the boolean `disableBackgroundThrottling` , [like electron does](https://www.electronjs.org/docs/latest/api/browser-window#:~:text=8859%2D1.-,backgroundThrottling,-boolean%20(optional)%20%2D%20Whether). And maybe add a `backgroundThrottlingPolicy` later that could be used for advanced cases. But then there'd be a dependency of the policy on the disable flag.
> 
> I would assume that almost no developer would decide for a semi-throttled/semi-unthrottled mode. In that case, they should rather focus on optimising the usage of timers in their code and keep it fully disabled.

Because Wry is an underlying library, I prefer to give higher-level libs/applications more flexibility in designing their API.
For example, we could add `disableBackgroundThrottling` in Tauri as a public API, and let tauri-wry-runtime decide which policy should be used.

pewsheen commented on src/lib.rs:
```diff
@@ -1181,6 +1197,26 @@ impl<'a> WebViewBuilder<'a> {
+  /// - **macOS**: Supported since version 14.0+.
+  ///
+  /// see https://github.com/tauri-apps/tauri/issues/5250#issuecomment-2569380578
+  pub fn with_background_throttling(self, policy: Option<BackgroundThrottlingPolicy>) -> Self {
+    self.and_then(|mut b| {
+      b.attrs.background_throttling = policy;
+      Ok(b)
+    })
+  }
```
```suggestion
  pub fn with_background_throttling(self, policy: BackgroundThrottlingPolicy) -> Self {
    self.and_then(|mut b| {
      b.attrs.background_throttling = Some(policy);
      Ok(b)
    })
  }
```

I assume the user calls `with_background_throttling` when they decide to set a non-default value, so `Option<>` may not be necessary here.

bastiankistner replied:
@pewsheen changed it accordingly, thank you!

bastiankistner committed:
chore: do not use an option for throttling policy

pewsheen commented:
@bastiankistner  Cool! Thanks for your contribution! ❤️

bastiankistner commented:
Thank you for your help!
